plutus-core/plutus-ir/src/PlutusIR/TypeCheck/Internal.hs

```diff
@@ -219,21 +220,28 @@ ty ~> vTy
 -}
 inferTypeM (Let ann r@NonRec bs inTerm) = do
     -- Check each binding individually, then if ok, introduce its new type/vars to the (linearly) next let or inTerm
-    ty <- foldr checkBindingThenScope (inferTypeM inTerm) bs
+    ty <- foldr checkBindingThenScope (inferTypeM (substTyBinds inTerm)) bs
     -- check the in-term's inferred type has kind * (except at toplevel)
     checkStarInferred ann ty
     pure ty
- where
-   checkBindingThenScope :: Binding TyName Name uni fun ann -> PirTCEnv uni fun e res -> PirTCEnv uni fun e res
-   checkBindingThenScope b acc = do
-       -- check that the kinds of the declared types are correct
-       checkKindFromBinding b
-       -- check that the types of declared terms are correct
-       checkTypeFromBinding r b
-       -- add new *normalized* termvariables to env
-       -- Note that the order of adding typesVSkinds here does not matter
-       withTyVarsOfBinding b $
-           withVarsOfBinding r b acc
+  where
+    checkBindingThenScope :: Binding TyName Name uni fun ann -> PirTCEnv uni fun e res -> PirTCEnv uni fun e res
+    checkBindingThenScope b acc = do
+        -- check that the kinds of the declared types are correct
+        checkKindFromBinding b
+        -- check that the types of declared terms are correct
+        checkTypeFromBinding r b
+        -- add new *normalized* termvariables to env
+        -- Note that the order of adding typesVSkinds here does not matter
+        withTyVarsOfBinding b $
+            withVarsOfBinding r b acc
+
+    substTyBinds = termSubstTyNames $ foldl' f (const Nothing) bs
```

Comment:
Once we're doing this on `ty` we'll need to use `substNormalizeType`, which retains the property that the resulting type is normalized (since we return normalized types from `inferTypeM`. You'll still want to collect all the types to substitute like this, obviously.

There will be a bit of duplicated work since you'll end up normalizing the RHSs of the type bindings twice, once here and once in the "main" part of the check, but I think that's okay for simplicity.